src/lib.rs

```diff
@@ -1561,27 +1559,30 @@ impl Builder {
     }
 
     /// Generate the Rust bindings using the options built up thus far.
-    pub fn generate(mut self) -> Result<Bindings, BindgenError> {
+    pub fn generate(self) -> Result<Bindings, BindgenError> {
+        let mut options = self.options.clone();
         // Add any extra arguments from the environment to the clang command line.
-        self.options.clang_args.extend(get_extra_clang_args());
+        options.clang_args.extend(get_extra_clang_args());
 
         // Transform input headers to arguments on the clang command line.
-        self.options.input_header = self.input_headers.pop();
-        self.options.extra_input_headers = self.input_headers;
-        self.options.clang_args.extend(
-            self.options.extra_input_headers.iter().flat_map(|header| {
-                iter::once("-include".into())
-                    .chain(iter::once(header.to_string()))
-            }),
+        options.clang_args.extend(
+            options.input_headers
+                [..options.input_headers.len().saturating_sub(1)]
+                .iter()
+                .flat_map(|header| ["-include".into(), header.to_string()]),
         );
 
-        let input_unsaved_files = self
-            .input_header_contents
-            .into_iter()
-            .map(|(name, contents)| clang::UnsavedFile::new(&name, &contents))
-            .collect::<Vec<_>>();
+        let input_unsaved_files =
+            std::mem::take(&mut options.input_header_contents)
+                .into_iter()
+                .map(|(name, contents)| clang::UnsavedFile::new(name, contents))
+                .collect::<Vec<_>>();
 
-        Bindings::generate(self.options, input_unsaved_files)
+        match Bindings::generate(options, input_unsaved_files) {
+            Ok(bindings) => Ok(bindings),
+            Err(GenerateError::ShouldRestart) => self.generate(),
```

Comment:
Why? I don't get this. This is dead code and would trivially loop right?

Reply:
actually let me do a change that could make this clearer.

Reply:
see https://github.com/rust-lang/rust-bindgen/pull/2292#issuecomment-1267282366